exporter/cortexexporter/cortex.go

```diff
@@ -0,0 +1,354 @@
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package cortexexporter
+
+import (
+	"bufio"
+	"bytes"
+	"context"
+	"fmt"
+	"io"
+	"net/http"
+	"strconv"
+	"strings"
+	"sync"
+
+	"github.com/gogo/protobuf/proto"
+	"github.com/golang/snappy"
+	"github.com/pkg/errors"
+	"github.com/prometheus/prometheus/prompb"
+
+	"go.opentelemetry.io/collector/consumer/pdata"
+	"go.opentelemetry.io/collector/consumer/pdatautil"
+	"go.opentelemetry.io/collector/internal/data"
+	otlp "go.opentelemetry.io/collector/internal/data/opentelemetry-proto-gen/metrics/v1"
+)
+
+// TODO: get default labels such as job or instance from Resource
+
+// cortexExporter converts OTLP metrics to Cortex TimeSeries and sends them to a remote endpoint
+type cortexExporter struct {
+	namespace string
+	endpoint  string
+	client    *http.Client
+	headers   map[string]string
+	wg        *sync.WaitGroup
+	closeChan chan struct{}
+}
+
+// newCortexExporter initializes a new cortexExporter instance and sets fields accordingly.
+// client parameter cannot be nil.
+func newCortexExporter(ns string, ep string, client *http.Client) (*cortexExporter, error) {
+
+	if client == nil {
+		return nil, fmt.Errorf("http client cannot be nil")
+	}
+
+	return &cortexExporter{
+		namespace: ns,
+		endpoint:  ep,
+		client:    client,
+		wg:        new(sync.WaitGroup),
+		closeChan: make(chan struct{}),
+	}, nil
+}
+
+// shutdown stops the exporter from accepting incoming calls(and return error), and wait for current export operations
+// to finish before returning
+func (ce *cortexExporter) shutdown(context.Context) error {
+	close(ce.closeChan)
+	ce.wg.Wait()
+	return nil
+}
+
+// pushMetrics converts metrics to Cortex TimeSeries and send to remote endpoint. It maintain a map of TimeSeries,
+// validates and handles each individual metric, adding the converted TimeSeries to the map, and finally
+// exports the map.
+func (ce *cortexExporter) pushMetrics(ctx context.Context, md pdata.Metrics) (int, error) {
+	ce.wg.Add(1)
+	defer ce.wg.Done()
+	select {
+	case <-ce.closeChan:
+		return pdatautil.MetricCount(md), fmt.Errorf("shutdown has been called")
+	default:
+		// stores TimeSeries
+		tsMap := map[string]*prompb.TimeSeries{}
+		dropped := 0
+		errs := []string{}
+
+		rms := data.MetricDataToOtlp(pdatautil.MetricsToInternalMetrics(md))
+
+		for _, r := range rms {
+			// TODO: add resource attributes as labels
+			for _, inst := range r.InstrumentationLibraryMetrics {
+				//TODO: add instrumentation library information as labels
+				for _, m := range inst.Metrics {
```

Comment:
I think all of these variable names (rms, r, inst, m) are pretty opaque- would prefer more clearly readable names